applications/OptimizationApplication/custom_utilities/response/mass_response_utils.cpp

```diff
@@ -0,0 +1,271 @@
+        //         return 3.0 * std::pow(ElementSizeCalculator<3, 8>::AverageElementSize(rGeometry), 2) * ElementSizeCalculator<3, 8>::AverageElementSizeDerivative(NodeIndex, DirectionIndex, rGeometry);
+        //     };
+        //     break;
+        default:
+            KRATOS_ERROR << "Non supported geometry type for mass shape sensitivity calculation (CalculateMassShapeSensitivity())." << std::endl;
+    }
+
+    block_for_each(rModelPart.Elements(), [&](auto& rElement){
+        auto& r_geometry = rElement.GetGeometry();
+        const IndexType dimension = r_geometry.WorkingSpaceDimension();
+
+        const double density = rElement.GetProperties()[DENSITY];
+        const double thickness = get_thickness(rElement);
+        const double cross_area = get_cross_area(rElement);
+
+        for (IndexType c = 0; c < r_geometry.PointsNumber(); ++c) {
+            auto& r_derivative_value = r_geometry[c].GetValue(rOutputSensitivityVariable);
+
+            for (IndexType k = 0; k < dimension; ++k) {
+                const double derivative_value = volume_derivative_method(c, k, r_geometry) * thickness * density * cross_area;
+                AtomicAdd(r_derivative_value[k], derivative_value);
+            }
+        }
+    });
+
+    rModelPart.GetCommunicator().AssembleNonHistoricalData(rOutputSensitivityVariable);
+
+    KRATOS_CATCH("")
+}
+
+void MassResponseUtils::CalculateMassDensitySensitivity(
+    ModelPart& rModelPart,
+    const Variable<double>& rOutputSensitivityVariable)
+{
+    KRATOS_TRY
+
+    KRATOS_ERROR_IF_NOT(HasVariableInProperties(rModelPart, DENSITY))
+        << "DENSITY is not found in element properties of " << rModelPart.FullName() << ".\n";
+
+    KRATOS_ERROR_IF(HasVariableInProperties(rModelPart, THICKNESS) && HasVariableInProperties(rModelPart, CROSS_AREA))
+        << rModelPart.FullName()
+        << " has elements with properties having both THICKNESS and CROSS_AREA. Please separate the model part such that either one of them is present in elemental properties.\n";
+
+    const auto get_thickness = HasVariableInProperties(rModelPart, THICKNESS)
+                                    ? [](const ModelPart::ElementType& rElement) { return rElement.GetProperties()[THICKNESS]; }
+                                    : [](const ModelPart::ElementType& rElement) { return 1.0; };
+
+    const auto get_cross_area = HasVariableInProperties(rModelPart, CROSS_AREA)
+                                    ? [](const ModelPart::ElementType& rElement) { return rElement.GetProperties()[CROSS_AREA]; }
+                                    : [](const ModelPart::ElementType& rElement) { return 1.0; };
+
+    block_for_each(rModelPart.Elements(), [&](auto& rElement) {
+        rElement.GetProperties().SetValue(rOutputSensitivityVariable, rElement.GetGeometry().DomainSize() * get_thickness(rElement) * get_cross_area(rElement));
+    });
+
+    KRATOS_CATCH("")
+}
+
+void MassResponseUtils::CalculateMassThicknessSensitivity(
```

Comment:
Definitely, we can simplify the implementation by creating a single function for sensitivity analysis. Based on the type, we can then call different sensitive functions. Furthermore, we need to be able to call a single function from Python to compute the sensitivity of a response with respect to different types. However, it's important to note that this implementation does not support asking for sensitivities with respect to multiple properties in the case of concurrent optimization. We will need to address this issue in our implementation. Thank you.

Reply:
Done :D